lxd-agent/api_1.0.go

```diff
@@ -61,3 +67,89 @@ func api10Get(d *Daemon, r *http.Request) response.Response {
 
 	return response.SyncResponseETag(true, fullSrv, fullSrv)
 }
+
+func api10Put(d *Daemon, r *http.Request) response.Response {
+	var data api.API10Put
+
+	err := json.NewDecoder(r.Body).Decode(&data)
+	if err != nil {
+		return response.ErrorResponse(http.StatusInternalServerError, err.Error())
+	}
+
+	d.serverCID = data.CID
+	d.serverPort = data.Port
+	d.serverCertificate = data.Certificate
+
+	// Try connecting to LXD server.
+	client, err := getClient(int(d.serverCID), int(d.serverPort), d.serverCertificate)
+	if err != nil {
+		return response.ErrorResponse(http.StatusInternalServerError, err.Error())
+	}
+
+	server, err := lxd.ConnectLXDHTTP(nil, client)
+	if err != nil {
+		return response.ErrorResponse(http.StatusInternalServerError, err.Error())
+	}
+
+	defer server.Disconnect()
+
+	// Let LXD know, we were able to connect successfully.
+	d.chConnected <- struct{}{}
+
+	if data.Devlxd {
+		err = startDevlxdServer(d)
```

Comment:
@monstermunchkin does `startDevlxdServer` need updating so that if a future PUT requests come in with data.Devlxd set to true then we don't end up starting two devlxd servers, or leaving go routines behind? Same question for `stopDevlxdServer`, can we call that function multiple times safely if it was previously stopped?

Reply:
For `startDevlxdServer` we could use a `devlxdRunning` variable which is guarded by a mutex. `stopDevlxdServer` can be called multiple times without a problem.

Reply:
Yeah I think calling `startDevlxdServer` for these scenarios needs to be handled/thought about (if not already):

- When the devlxd server is stopping.
- When the devlxd server is stopped.
- When the devlxd server is starting.
- When the devlxd server is started.

As you say this will likely need use of a mutex inside those functions.